### ISSUE TITLE
Problem with Read 2016 results

### ISSUE THREAD

**Issue by dparres** [FactoDeepLearning/DAN#9]
Hello!

I am trying to replicate the results for READ 2016. The pre-train works very well, and I get the following results:

EPOCH 1667/10000: 100%|███████████████████████████████████████████████████████████████████████| 8367/8367 [00:59<00:00, 140.63it/s, values={'loss_ctc': 0.0133, 'cer': 0.0031, 'wer': 0.0102}]
LOADED EPOCH: 1488

Prediction: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████| 1140/1140 [00:09<00:00, 118.05it/s, values={'cer': 0.0001, 'wer': 0.0009}]
Prediction: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████| 1043/1043 [00:08<00:00, 119.96it/s, values={'cer': 0.0, 'wer': 0.0005}]
Prediction: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████| 8367/8367 [01:12<00:00, 114.81it/s, values={'cer': 0.0, 'wer': 0.0}]

But in training, the model does not converge, and the cer does not go below 70%. How can I replicate the results correctly? I have been increasing the number of steps before increasing the number of synthetic lines, and I can't achieve competitive results. These are the logs I get:

EPOCH 66/50000: 100%|███████████████████████████████████████████████████████| 350/350 [00:26<00:00, 13.09it/s, values={'loss_ce': 0.3353, 'cer': 0.7057, 'wer': 0.9796, 'syn_max_lines': 1.0}]
EPOCH 67/50000: 100%|████████████████████████████████████████████████████████| 350/350 [00:26<00:00, 13.19it/s, values={'loss_ce': 0.3393, 'cer': 0.7091, 'wer': 0.997, 'syn_max_lines': 1.0}]
EPOCH 68/50000: 100%|███████████████████████████████████████████████████████| 350/350 [00:27<00:00, 12.96it/s, values={'loss_ce': 0.3278, 'cer': 0.7041, 'wer': 0.9851, 'syn_max_lines': 1.0}]
EPOCH 69/50000: 100%|███████████████████████████████████████████████████████| 350/350 [00:28<00:00, 12.33it/s, values={'loss_ce': 0.3742, 'cer': 0.7081, 'wer': 1.0143, 'syn_max_lines': 1.0}]
EPOCH 70/50000: 100%|███████████████████████████████████████████████████████| 350/350 [00:28<00:00, 12.12it/s, values={'loss_ce': 0.4485, 'cer': 0.7125, 'wer': 0.9969, 'syn_max_lines': 1.0}]
Evaluation E70: 100%|██████████████████████████████████████████████████████████████████████████████| 50/50 [11:01<00:00, 13.23s/it, values={'cer': 6.5951, 'wer': 11.0707, 'map_cer': 0.0063}]
EPOCH 71/50000: 100%|███████████████████████████████████████████████████████| 350/350 [00:27<00:00, 12.94it/s, values={'loss_ce': 0.3595, 'cer': 0.7111, 'wer': 0.9975, 'syn_max_lines': 1.0}]
EPOCH 72/50000: 100%|███████████████████████████████████████████████████████| 350/350 [00:27<00:00, 12.50it/s, values={'loss_ce': 1.6635, 'cer': 0.7639, 'wer': 1.1048, 'syn_max_lines': 2.0}]
EPOCH 73/50000: 100%|████████████████████████████████████████████████████████| 350/350 [00:28<00:00, 12.36it/s, values={'loss_ce': 1.2317, 'cer': 0.734, 'wer': 1.0037, 'syn_max_lines': 2.0}]
EPOCH 74/50000: 100%|███████████████████████████████████████████████████████| 350/350 [00:27<00:00, 12.79it/s, values={'loss_ce': 1.1756, 'cer': 0.7231, 'wer': 1.0299, 'syn_max_lines': 2.0}]
EPOCH 75/50000: 100%|████████████████████████████████████████████████████████| 350/350 [00:27<00:00, 12.54it/s, values={'loss_ce': 1.0596, 'cer': 0.7141, 'wer': 1.052, 'syn_max_lines': 2.0}]

Greetings and thank you very much,

Daniel

**Comment by FactoDeepLearning** [comment]
Hi, are you sure the weights from the line-level model are used as transfer learning ?
The number of steps before increasing the number of synthetic lines does not need to be changed

**Comment by dparres** [comment]
I am loading the weights correctly, and I even tested with your "fcn_read_2016_line_syn.pt". Also, I have checked that "adapt_decision_layer_to_old_charset" is applied correctly.

But I found that during training, if init_proba=0.9, the model gets the results I reported above and never learned. But if I change init_proba=1. The model learns to transcribe until the number of synth lines is increased.

![image](https://user-images.githubusercontent.com/114649578/210232483-ec44fe2d-d6d5-4182-b5a0-5f80829d1ff6.png)

I have tried several times to do the DAN training for documents in READ 2016, and I can't get good results. I am performing the GitHub steps. Is there something I might be missing?

Best regards and thank you very much,

Daniel

**Comment by FactoDeepLearning** [comment]
Do you have the "transfered weights for encoder" print before training begins ?
Can you share the "params" file in the output folder, to check that everything is fine.

**Comment by dparres** [comment]
Yes, that print appears. I put another log to check the path and attached a screenshot. I am sharing the "params" I have converted it to ".txt" so that GitHub allows me to share it.

![image](https://user-images.githubusercontent.com/114649578/210233952-e404400c-17b9-4416-a6c6-5f497e69bff9.png)

[params.txt](https://github.com/FactoDeepLearning/DAN/files/10331684/params.txt)

Thank you very much,

Daniel

**Comment by FactoDeepLearning** [comment]
At first sight, it seems that there is no config issues. In case it has an impact, here is the software versions I used:  "pytorch_version": "1.8.2", "cuda_version": "10.2", "cudnn_version": 7605.

Otherwise, you can check that the generated images are correct in the "train_batch" function (trainer_dan.py) by comparing image (self.dataset.train_dataset.denormalize(batch_data["imgs_shape"])) and associated label (batch_data["raw_labels"]). 

In any case, convergence is slow, you can try waiting for more epochs (2/3 thousands for full training)

**Comment by dparres** [comment]
OK! Do you have any logs of when you trained DAN for READ 2016? This would be a great help in replicating the experiment.
Best regards and thank you very much,
Daniel

**Comment by FactoDeepLearning** [comment]
Here is what is expected for READ 2016

Training CER:
![training](https://user-images.githubusercontent.com/25418418/210241636-a24d05f7-0928-4956-b91b-0f721fdda20b.png)

Validation CER:
(quite messy at first, because we mainly train on few lines only whereas validation examples are whole documents)
![validation](https://user-images.githubusercontent.com/25418418/210241638-9c4caff2-5824-4c40-8394-9213de8f4547.png)

**Comment by dparres** [comment]
Thank you very much! That's what I needed! :)
